File: c/Visitor_Shop.c
/// @file
/// @brief
/// Implementation of the Visitor_Shop structure's supporting functions,
/// Shop_Create(), Shop_Destroy(), Shop_PlaceOrder(), and Shop_PickupOrder(),
/// as used in the @ref visitor_pattern.

#include <stdio.h>
#include <memory.h>

#include "helpers/dynamicstring.h"

#include "Visitor_Village.h"

/// <summary>
/// Convert a string list to a comma-delimited string.  Useful for
/// displaying the list.
/// </summary>
/// <param name="list">The ConstStringList object containing the strings to be
/// stringized</param>
/// <param name="output">A DynamicString object that returns the resulting
/// string.</param>
/// <returns>Returns true if the operation completed successfully; otherwise,
/// returns false, indicating an out of memory condition (or a NULL argument).
/// </returns>
static bool _StringizeStringList(ConstStringList* list, DynamicString* output)
{
    bool success = false;

    if (list != NULL && output != NULL)
    {
        for (size_t index = 0; index < list->strings_count; index++)
        {
            if (index != 0)
            {
                success = DynamicString_Append(output, ", ");
                if (!success)
                {
                    printf("  Error!  Out of memory condition appending ',' to list of strings being concatenated into a single string!\n");
                    break;
                }
            }
            success = DynamicString_Append(output, list->strings[index]);
            if (!success)
            {
                printf("  Error!  Out of memory condition appending a string to list of strings being concatenated into a single string!\n");
                break;
            }
        }
    }

    return success;
}

/// <summary>
/// Determine if the given Visitor_Shop object sells the specified item.
/// </summary>
/// <param name="shop">The Visitor_Shop to access.</param>
/// <param name="item">The item to look for</param>
/// <returns>Returns true if this shop sells the item; otherwise false.
/// </returns>
static bool Shop_DoesShopSellItem(Visitor_Shop* shop, const char* item)
{
    bool doesSell = false;

    if (shop != NULL)
    {
        int foundIndex = MapOfStrings_Find(&shop->IngredientsForItems, item);
        doesSell = foundIndex != -1;
    }

    return doesSell;
}


/// <summary>
/// Determine if given Visitor_Shop object has the specified item in stock.
/// </summary>
/// <param name="shop">The Visitor_Shop to access.</param>
/// <param name="item">The item to look for</param>
/// <returns>Returns true if the item is in stock; otherwise false.
/// </returns>
static bool Shop_IsItemInStock(Visitor_Shop* shop, const char* item)
{
    bool inStock = false;

    if (shop != NULL && item != NULL)
    {
        int foundIndex = MapOfInt_Find(&shop->Inventory, item);
        if (foundIndex != -1)
        {
            inStock = shop->Inventory.entries[foundIndex].value > 0;
        }
    }

    return inStock;
}


/// <summary>
/// Add the specified item to the given Visitor_Shop object's inventory.
/// </summary>
/// <param name="shop">The Visitor_Shop to access.</param>
/// <param name="item">The item to add</param>
/// <returns>Returns true if the item was added successfully; otherwise,
/// returns false, indicating an out of memory condition (or a NULL argument).
/// </returns>
static bool Shop_AddItemToInventory(Visitor_Shop* shop, const char* item)
{
    bool success = false;

    if (shop != NULL && item != NULL)
    {
        int foundIndex = MapOfInt_Find(&shop->Inventory, item);
        if (foundIndex != -1)
        {
            shop->Inventory.entries[foundIndex].value++;
            success = true;
        }
        else
        {
            success = MapOfInt_Add(&shop->Inventory, item, 1);
            if (!success)
            {
                printf("  Error!  Out of memory condition adding an item to a shop's inventory!\n");
            }
        }
    }

    return success;
}


//-----------------------------------------------------------------------------
//-----------------------------------------------------------------------------


///////////////////////////////////////////////////////////////////////////////
// Shop_Create()
///////////////////////////////////////////////////////////////////////////////
Visitor_Shop* Shop_Create(const char* name, const char* address, struct Village* village)
{
    Visitor_Shop* shop = NULL;

    if (name != NULL && address != NULL && village != NULL)
    {
        shop = calloc(1, sizeof(Visitor_Shop));
        if (shop != NULL)
        {
            shop->Name = name;
            shop->Address = address;
            shop->Village = village;
            MapOfStrings_Initialize(&shop->IngredientsForItems);
            MapOfInt_Initialize(&shop->Inventory);
        }
    }

    return shop;
}

///////////////////////////////////////////////////////////////////////////////
// Shop_Destroy()
///////////////////////////////////////////////////////////////////////////////
void Shop_Destroy(Visitor_Shop* shop)
{
    if (shop != NULL)
    {
        MapOfStrings_Clear(&shop->IngredientsForItems);
        MapOfInt_Clear(&shop->Inventory);
    }
}



///////////////////////////////////////////////////////////////////////////////
// Shop_PlaceOrder()
///////////////////////////////////////////////////////////////////////////////
PlaceOrderReponse Shop_PlaceOrder(Visitor_Shop* shop, ConstStringList* itemsToOrder)
{
    PlaceOrderReponse response = PlaceOrderResponse_Error;

    if (itemsToOrder != NULL && shop != NULL)
    {
        ConstStringList outOfStockItems = { 0 };
        ConstStringList itemsInThisShop = { 0 };

        bool success = false;

        for (size_t index = 0; index < itemsToOrder->strings_count; index++)
        {
            const char* item = itemsToOrder->strings[index];
            if (Shop_DoesShopSellItem(shop, item))
            {
                if (!Shop_IsItemInStock(shop, item))
                {
                    success = ConstStringList_AddString(&outOfStockItems, item);
                    if (!success)
                    {
                        printf("  Error!  Out of memory condition adding ingredient to out of stock item list!\n");
                        break;
                    }
                }
                success = ConstStringList_AddString(&itemsInThisShop, item);
                if (!success)
                {
                    printf("  Error!  Out of memory condition adding ingredient to local list of items sold by a store!\n");
                    break;
                }
            }
        }

        if (success)
        {
            if (itemsInThisShop.strings_count != 0)
            {
                DynamicString output = { 0 };
                success = _StringizeStringList(&itemsInThisShop, &output);
                if (success)
                {
                    printf("  %s: Received an order for %s.\n",
                        shop->Name, output.string);
                    response = PlaceOrderResponse_OrderAccepted;
                }
                DynamicString_Clear(&output);
            }
            else
            {
                response = PlaceOrderResponse_OrderIgnored;
            }

        }

        if (success)
        {
            if (outOfStockItems.strings_count != 0)
            {
                for (size_t index = 0; index < outOfStockItems.strings_count; index++)
                {
                    const char* ingredient = outOfStockItems.strings[index];
                    int ingredientIndex = MapOfStrings_Find(&shop->IngredientsForItems, ingredient);
                    if (ingredientIndex != -1 && shop->IngredientsForItems.entries[ingredientIndex].value->strings_count != 0)
                    {
                        printf("  %s:   %s out of stock, ordering ingredients to make more...\n",
                                shop->Name, ingredient);
                        OrderVisitor visitor = { 0 };
                        success = ConstStringList_AddStrings(&visitor.ItemsToOrder, shop->IngredientsForItems.entries[ingredientIndex].value->strings, shop->IngredientsForItems.entries[ingredientIndex].value->strings_count);
                        if (!success)
                        {
                            printf("  Error!  Out of memory creating order for out of stock items!\n");
                            break;
                        }
                        success = Village_VisitShop(shop->Village, &visitor);
                        if (!success)
                        {
                            printf("  Error! Failed to visit shops for out of stock items, probably an out of memory condition!\n");
                            break;
                        }
                    }
                    else
                    {
                        // The ordered item has no ingredients so the
                        // ordered item will be magically added to inventory
                        printf("  %s:   %s out of stock, making...\n",
                                shop->Name, ingredient);
                    }
                    success = Shop_AddItemToInventory(shop, ingredient);
                    if (!success)
                    {
                        break;
                    }
                }
            }
        }
        if (!success)
        {
            response = PlaceOrderResponse_Error;
        }

        ConstStringList_Clear(&outOfStockItems);
        ConstStringList_Clear(&itemsInThisShop);
    }

    return response;
}


///////////////////////////////////////////////////////////////////////////////
// Shop_PickupOrder()
///////////////////////////////////////////////////////////////////////////////
bool Shop_PickupOrder(Visitor_Shop* shop, ConstStringList* items, ConstStringList* itemsToBePickedUp)
{
    bool success = false;

    if (items != NULL && itemsToBePickedUp  != NULL && shop != NULL)
    {
        ConstStringList_Clear(itemsToBePickedUp);
        for (size_t index = 0; index < items->strings_count; index++)
        {
            const char* item = items->strings[index];
            // If this shop sells the item and the item is in stock then
            if (Shop_DoesShopSellItem(shop, item))
            {
                if (Shop_IsItemInStock(shop, item))
                {
                    success = ConstStringList_AddString(itemsToBePickedUp, item);
                    if (!success)
                    {
                        printf("  Error!  Out of memory adding item to list of items to be picked up!\n");
                        break;
                    }
                }
                else
                {
                    printf("  Error!  %s: Item %s is not in the inventory when it should be.\n",
                            shop->Name, item);
                }
            }
        }

        if (itemsToBePickedUp->strings_count != 0)
        {
            // Reduce inventory for the ordered items
            DynamicString output = { 0 };
            success = _StringizeStringList(itemsToBePickedUp, &output);
            if (success)
            {
                for (size_t index = 0; index < itemsToBePickedUp->strings_count; index++)
                {
                    const char* itemToBePickedUp = itemsToBePickedUp->strings[index];
                    int inventoryIndex = MapOfInt_Find(&shop->Inventory, itemToBePickedUp);
                    if (inventoryIndex != -1)
                    {
                        shop->Inventory.entries[inventoryIndex].value--;
                    }
                }
                printf("  %s: Order picked up for %s.\n", shop->Name, output.string);
            }
            DynamicString_Clear(&output);
        }

    }
    return success;
}

File: rust/design_pattern_examples_rust/src/visitor/visitor_shop.rs
//! Contains the VisitorShop struct, which represents each shop in the village
//! that can be visited.

//-----------------------------------------------------------------------------

use std::collections::HashMap;

use super::visitor_ordervisitor::OrderVisitor;


//-----------------------------------------------------------------------------

/// Convert a string list to a comma-delimited string.  Useful for displaying
/// the list.
///
/// # Parameters
/// - items
///
///   List of strings to be formatted into a comma-delimited string.
///
/// # Returns
/// Returns a string containing the formatted list.
fn stringize_list(items: &Vec<String>) -> String {
    items.join(", ")
}
//-----------------------------------------------------------------------------


pub struct VisitorShop {
    /// Name of this shop.
    pub name: String,
    /// Address of this shop.
    pub address: String,
    //struct Village* Village : Can't use circular references!
    /// Maps ingredient to list of items needed to make ingredient.
    pub ingredients_for_items: HashMap<String, Vec<String>>,
    /// Maps ingredient to count of that ingredient currently in the shop.
    pub inventory: HashMap<String, i32>,
}

impl VisitorShop {
    /// Constructor.
    ///
    /// # Parameters
    /// - name
    ///
    ///   Name of the shop.
    /// - address
    ///
    ///   Address of the shop.
    /// - ingredients_for_items
    ///
    ///   Map of items to be sold with a list of ingredients needed to "make"
    ///   each item.  The VisitorShop takes ownership of the map.
    ///
    /// # Returns
    /// Returns a new instance of the VisitorShop struct.
    pub fn new(name: &str, address: &str, ingredients_for_items: HashMap<String, Vec<String>>) -> VisitorShop {
        let mut inventory: HashMap<String, i32> = HashMap::new();
        for item in ingredients_for_items.keys() {
            inventory.insert(item.clone(), 0);
        }
        VisitorShop {
            name: name.to_string(),
            address: address.to_string(),
            ingredients_for_items,
            inventory: inventory,
        }
    }

    /// Determine if this shop sells the specified item.
    ///
    /// # Parameters
    /// - item
    ///
    ///   The name of the item to look for.
    ///
    /// # Returns
    /// Returns true if this shop sells the item; otherwise, returns false.
    fn does_shop_sell_item(&self, item: &str) -> bool {
        self.ingredients_for_items.contains_key(item)
    }

    /// Determine if this shop has the specified item in stock.
    ///
    /// # Parameters
    /// - item
    ///
    ///   The name of the item to look for.
    ///
    /// # Returns
    /// Returns true if the item is in stock; otherwise, returns false.
    fn is_item_in_stock(&self, item: &str) -> bool {
        if self.inventory.contains_key(item) {
            self.inventory[item] > 0
        } else {
            false
        }
    }

    /// Add the specified item to this shop's inventory.
    ///
    /// # Parameters
    /// - item
    ///
    ///   The name of the item to update its inventory count.
    fn add_item_to_inventory(&mut self, item: &str) {
        if !self.inventory.contains_key(item) {
            self.inventory.insert(item.to_string(), 1);
        } else {
            self.inventory.entry(item.to_string()).and_modify(|e| { *e += 1 });
        }
    }

    /// Visit this shop to try to place an order as described in the
    /// OrderVisitor object.
    ///
    /// # Parameters
    /// - order
    ///
    ///   An OrderVisitor object that contains a list of items to order,
    ///   possibly at this shop.
    ///
    /// # Returns
    /// Returns true if the order was successfully placed at this shop;
    /// otherwise, returns false, try another shop.
    pub fn place_order(&mut self, order: &mut OrderVisitor) -> bool {
        let mut order_placed = false;
        
        let mut out_of_stock_items: Vec<String> = vec![];
        let mut items_in_this_shop: Vec<String> = vec![];
        
        for item in order.items_to_order.iter() {
            if self.does_shop_sell_item(&item) {
                if !self.is_item_in_stock(&item) {
                    out_of_stock_items.push(item.clone());
                }
                items_in_this_shop.push(item.clone());
            }
        }

        if !items_in_this_shop.is_empty() {
            println!("  {0}: Received an order for {1}.",
                    self.name, stringize_list(&items_in_this_shop));
            order_placed = true;
        }

        if !out_of_stock_items.is_empty() {
            for item in out_of_stock_items.iter() {
                let ingredients = self.ingredients_for_items.get(item);
                if let Some(items) = ingredients {
                    if !items.is_empty() {
                        println!("  {0}:   {1} out of stock, ordering ingredients to make more...",
                            self.name, item);
                        let reorder_visitor = OrderVisitor::new(&items);
                        //village.visit(&reorder_visitor);
                    } else {
                        // The ordered item has no ingredients so the
                        // ordered item will be magically added to inventory
                        println!("  {0}:   {1} out of stock, making...",
                                self.name, item);
                        self.add_item_to_inventory(item);
                    }
                }
            }
        }

        order_placed
    }


    /// Pick up the items sold by this shop (assumes the items were ordered
    /// already).  Basically, this reduces the inventory for the given items
    /// that are sold by this shop.  This method should be called only if
    /// place_order() returns true.
    ///
    /// # Parameters
    /// - order
    ///
    ///   An OrderVisitor object that receives the requested items.
    pub fn pickup_order(&mut self, order: &mut OrderVisitor) {
        order.items_received.clear();
        for item in order.items_to_order.iter() {
            if self.does_shop_sell_item(&item) {
                if self.is_item_in_stock(&item) {
                    order.items_received.push(item.clone());
                } else {
                    println!("  Error!  {0}: Item {1} is not in the inventory when it should be.",
                            self.name, item);
                }
            }
        }
        if !order.items_received.is_empty() {
            // Reduce inventory for the ordered items
            let items_as_string = stringize_list(&order.items_received);
            for item in order.items_received.iter() {
                self.inventory.entry(item.clone()).and_modify(|e| { *e -= 1 });
            }
            println!("  {0}: Order picked up for {1}.", self.name, items_as_string);
        }
    }
}
